File: src/jkarabo/karabo/util/Configurator.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package karabo.util;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;
import static karabo.util.ClassInfo.inheritanceChain;
import karabo.util.vectors.VectorHash;
import karabo.util.vectors.VectorString;

/**
 * 
 * @author Sergey Esenov serguei.essenov at xfel.eu
 * @param <Base> anything that inherits from Object
 */
@KARABO_CLASSINFO(classId = "Configurator", version = "1.0")
public class Configurator<Base> {

    private final Map<String, Map<String, Class<?>>> m_registry = new LinkedHashMap<>();
    private static String m_default;
    private Class<Base> base;
    private static Map<Class, Configurator> configurators = new LinkedHashMap<>();

    private static <BaseClass> Configurator init(Class<BaseClass> base) {
        if (configurators.get(base) == null) {
            configurators.put(base, new Configurator<>(base));
        }
        return configurators.get(base);
    }

    private Configurator(Class<Base> base) {
        this.base = base;
    }

    private static <B> String ctorKey(Class<B> base) {
        return base.getName() + "_" + Hash.class.getName();
    }

    private <B, A1> String ctorKey(Class<B> base, Class<A1> c1) {
        String h = base.getName() + "_" + Hash.class.getName();
        String a1 = c1.getName();
        return h + "," + a1;
    }

    private <B, A1, A2> String ctorKey(Class<B> base, Class<A1> c1, Class<A2> c2) {
        String h = base.getName() + "_" + Hash.class.getName();
        String a1 = c1.getName();
        String a2 = c2.getName();
        return h + "," + a1 + "," + a2;
    }

    private <B, A1, A2, A3> String ctorKey(Class<B> base, Class<A1> c1, Class<A2> c2, Class<A3> c3) {
        String h = base.getName() + "_" + Hash.class.getName();
        String a1 = c1.getName();
        String a2 = c2.getName();
        String a3 = c3.getName();
        return h + "," + a1 + "," + a2 + "," + a3;
    }

    public static <BaseClass, DerivedClass> void registerClass(Class<BaseClass> base, Class<DerivedClass> derived, String derivedClassId) {
        Configurator configurator = Configurator.init(base);
        if (!configurator.m_registry.containsKey(derivedClassId)) {
            configurator.m_registry.put(derivedClassId, new LinkedHashMap());
        }
        Map<String, Class<?>> r = (Map) configurator.m_registry.get(derivedClassId);
        String ctorStr = ctorKey(base);
        if (!r.containsKey(ctorStr)) {
            r.put(ctorStr, derived);
            //System.out.println("*** registerClass: classId is \"" + derivedClassId + "\" and ctor string is \"" + ctorStr + "\" -> class " + derived.getName());
        }
    }

    public void setDefault(String classId) {
        Configurator.m_default = classId;
    }

    public static Schema getSchema(Class base, String classId, Schema.AssemblyRules rules) {
        Configurator configurator = Configurator.init(base);
        Schema schema = new Schema(classId, rules);
        //System.out.println("Schema getSchema(Class base, String classId = " + classId + ", Schema.AssemblyRules rules)");
        if (configurator.m_registry.containsKey(classId)) {
            Map<String, Class<?>> r = (Map<String, Class<?>>) configurator.m_registry.get(classId);
            String ctorStr = (String)ctorKey(base);
            if (r.containsKey(ctorStr)) {
                Class<?> classObj = r.get(ctorStr);
                ArrayList<Class<?>> chain = inheritanceChain(new ArrayList<Class<?>>(), classObj, base);
                //System.out.println("getSchema inheritance chain : " + chain);
                for (Class<?> cls : chain) {
                    try {
                        Method m = cls.getDeclaredMethod("expectedParameters", new Class[]{Schema.class});
                        m.invoke(null, new Object[]{schema});
                    } catch (NoSuchMethodException ex) {
                    } catch (SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException ex) {
                        Logger.getLogger(Configurator.class.getName()).log(Level.SEVERE, null, ex);
                    }
                }
            } else {
                Logger.getLogger(Configurator.class.getName()).log(Level.WARNING,
                        "Subregistry for \"{0}\" does not contain ctor string \"{1}\".  Use KARABO_REGISTER_FOR_CONFIGURATION first.",
                        new Object[]{classId, ctorStr});
                return null;
            }
        } else {
            Logger.getLogger(Configurator.class.getName()).log(Level.WARNING,
                    "Registry does not contain \"{0}\".  Use KARABO_REGISTER_FOR_CONFIGURATION first.", classId);
            return null;
        }
        return schema;
    }

    public static Schema getSchema(Class base, String classId) {
        Configurator configurator = Configurator.init(base);
        Schema schema = new Schema(classId);
        //System.out.println("Schema getSchema(Class base, String classId = " + classId + ")");
        if (configurator.m_registry.containsKey(classId)) {
            Map<String, Class<?>> r = (Map<String, Class<?>>) configurator.m_registry.get(classId);
            String ctorStr = ctorKey(base);
            if (r.containsKey(ctorStr)) {
                Class<?> classObj = r.get(ctorStr);
                ArrayList<Class<?>> chain = inheritanceChain(new ArrayList<Class<?>>(), classObj, base);
                //System.out.println("getSchema inheritance chain : " + chain);
                for (Class<?> cls : chain) {
                    try {
                        Method m = cls.getDeclaredMethod("expectedParameters", new Class[]{Schema.class});
                        m.invoke(null, new Object[]{schema});
                    } catch (NoSuchMethodException ex) {
                    } catch (SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException ex) {
                        Logger.getLogger(Configurator.class.getName()).log(Level.SEVERE, null, ex);
                    }
                }
            } else {
                Logger.getLogger(Configurator.class.getName()).log(Level.WARNING,
                        "Subregistry for \"{0}\" does not contain ctor string \"{1}\".  Use KARABO_REGISTER_FOR_CONFIGURATION first.",
                        new Object[]{classId, ctorStr});
                return null;
            }
        } else {
            Logger.getLogger(Configurator.class.getName()).log(Level.WARNING,
                    "Registry does not contain \"{0}\".  Use KARABO_REGISTER_FOR_CONFIGURATION first.", classId);
            return null;
        }
        return schema;
    }

    public static void validateConfiguration(Class base, String classId, Hash configuration, Hash validated) {
        Schema schema = getSchema(base, classId);
        Validator validator = new Validator(); // Default validation
        ValidatorResult ret = validator.validate(schema, configuration, validated);
        if (!ret.success) {
            throw new RuntimeException("Validation failed. \n" + ret.message);
        }
    }

    public static <Base> Base createDefault(Class<Base> base, boolean validate) {
        Configurator configurator = Configurator.init(base);
        String defaultClassId = Configurator.m_default;
        if (defaultClassId.isEmpty()) {
            throw new RuntimeException("No default was defined");
        }
        return create(base, defaultClassId, new Hash(), validate);
    }

    public static <Base> Base createDefault(Class<Base> base) {
        return createDefault(base, true);
    }

    public static <Base, Derived extends Base> Base create(Class<Base> base, String classId, Hash configuration, boolean validate) {
        Configurator configurator = Configurator.init(base);
        if (!configurator.m_registry.containsKey(classId)) {
            throw new RuntimeException("ClassId: '" + classId + "' is not registered.");
        }
        Map<String, Class<?>> r = (Map<String, Class<?>>) configurator.m_registry.get(classId);
        String ctorStr = ctorKey(base);
        if (!r.containsKey(ctorStr)) {
            throw new RuntimeException("Ctor string: '" + ctorStr + "' is not registered in base registry '" + classId + "'.");
        }

        Class<?> derived = r.get(ctorStr);

//        System.out.println("*** Configurator: classId is \"" + classId + "\", ctor string is \""
//                + ctorStr + "\", Base class is \"" + base.getName() + "\", Derived class is \"" + derived.getName() + "\"");
        Constructor ctor;
        try {
            ctor = derived.getDeclaredConstructor(Hash.class);
        } catch (NoSuchMethodException | SecurityException ex) {
            Logger.getLogger(Configurator.class.getName()).log(Level.SEVERE, null, ex);
            return null;
        }

        try {
            if (validate) {
                Hash validated = new Hash();
                validateConfiguration(base, classId, configuration, validated);
                return (Base) ctor.newInstance(validated);
            } else {
                return (Base) ctor.newInstance(configuration);
            }
        } catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException ex) {
            Logger.getLogger(Configurator.class.getName()).log(Level.SEVERE, null, ex);
            return null;
        }
    }

    public static <Base, Derived extends Base> Base create(Class<Base> base, String classId, Hash configuration) {
        return create(base, classId, configuration, true);
    }

    public static <Base, Derived extends Base> Base create(Class<Base> base, String classId) {
        return create(base, classId, new Hash());
    }

    public static KeyValue<String, Hash> splitIntoClassIdAndConfiguration(Hash rootedConfiguration) {
        if (rootedConfiguration.size() != 1) {
            throw new RuntimeException("Expecting exactly one (root-)node identifying the classId in configuration");
        }
        Node node = rootedConfiguration.iterator().next().getValue();
        String classid = node.getKey();
        Hash config = (Hash) node.getValue();
        return new KeyValue<>(classid, config);
    }

    public static <Base> Base create(Class<Base> base, Hash configuration, boolean validate) {
        try {
            KeyValue<String, Hash> p = splitIntoClassIdAndConfiguration(configuration);
            return create(base, p.key, p.value, validate);
        } catch (Exception e) {
            throw new RuntimeException("This create method expects a rooted Hash with the root node name specifying the classId");
        }
    }

    public static <Base> Base create(Class<Base> base, Hash configuration) {
        return create(base, configuration, true);
    }

    public static <Base> Base createNode(Class<Base> base, String nodeName, String classId, Hash input, boolean validate) {
        if (input.has(nodeName)) {
            return create(base, classId, (Hash) input.get(nodeName), validate);
        } else {
            throw new RuntimeException("Given nodeName \"" + nodeName + "\" is not part of input configuration");
        }
    }

    public static <Base> Base createNode(Class<Base> base, String nodeName, String classId, Hash input) {
        return createNode(base, nodeName, classId, input, true);
    }

    public static <Base> Base createChoice(Class<Base> base, String choiceName, Hash input, boolean validate) {
        if (input.has(choiceName)) {
            return create(base, (Hash) input.get(choiceName), validate);
        } else {
            throw new RuntimeException("Given choiceName \"" + choiceName + "\" is not part of input configuration");
        }
    }

    public static <Base> Base createChoice(Class<Base> base, String choiceName, Hash input) {
        return createChoice(base, choiceName, input, true);
    }

    public static <Base> ArrayList<Base> createList(Class<Base> base, String listName, Hash input, boolean validate) {
        if (input.has(listName)) {
            VectorHash tmp = input.get(listName);
            ArrayList<Base> instances = new ArrayList<>();
            for (int i = 0; i < tmp.size(); ++i) {
                instances.add(create(base, tmp.get(i), validate));
            }
            return instances;
        } else {
            throw new RuntimeException("Given listName \"" + listName + "\" is not part of input configuration");
        }
    }

    public static <Base> ArrayList<Base> createList(Class<Base> base, String listName, Hash input) {
        return createList(base, listName, input, true);
    }

    public static <B> VectorString getRegisteredClasses(Class<B> base) {
        Configurator configurator = Configurator.init(base);
        VectorString registeredClasses = new VectorString();
        for (Map.Entry<String, Map<String, Class<?>>> entry : (Set<Entry<String, Map<String, Class<?>>>>) configurator.m_registry.entrySet()) {
            registeredClasses.add(entry.getKey());
        }
        return registeredClasses;
    }
}

File: src/pythonKarabo/karabo_package/configurator.py
# To change this template, choose Tools | Templates
# and open the template in the editor.

__author__="Sergey Esenov <serguei.essenov at xfel.eu>"
__date__ ="$Apr 11, 2013 4:20:13 PM$"

from karabo.karathon import Hash, Schema, AssemblyRules, AccessType, READ, WRITE, INIT, Validator

class Configurator(object):
    '''
    Configurator is the singleton class that keeps methods for registration and creation other classes that
    support "configuration" protocol, i.e. are configurable classes
    '''
    _instance = None
    registry = {}   # { classid_of_base_class : registry_for_classes_derived_of_the_base_class }
    
    @staticmethod
    def registerAsBaseClass(theClass):
        '''
        if hasattr(theClass, "__classid__"):
            print "registerAsBaseClass: class '{}' has __classid__ = {}".format(theClass.__name__, theClass.__classid__)
        else:
            print "registerAsBaseClass: class '{}' has no __classid__ attribute".format(theClass.__name__)
        if hasattr(theClass, "__base_classid__"):
            print "registerAsBaseClass: class '{}' has __base_classid__ = {}".format(theClass.__name__, theClass.__base_classid__)
        else:
            print "registerAsBaseClass: class '{}' has no __base_classid__ attribute".format(theClass.__name__)
        print "Class '{}.__classid__' found in registry: {}".format(theClass.__name__, theClass.__classid__ in Configurator.registry)
        '''
        if theClass.__classid__ not in Configurator.registry:
            theClass.__base_classid__ = theClass.__classid__
            theClass.__bases_classid__ = (theClass.__classid__,)
            Configurator.registry[theClass.__classid__] = {}
        Configurator.registry[theClass.__classid__][theClass.__classid__] = theClass  # self-registering

    
    def __init__(self, classid):
        '''
        The argument to the constructor may be the classid of a configurable class or configurable class itself:
        Example:
                c = Configurator(ConfigurableClass)
        or
                c = Configurator("ConfigurableClassId")
        '''
        if isinstance(classid, type):
            classid = classid.__classid__
        if not isinstance(classid, basestring):
            raise TypeError("The argument type '{}' is not allowed. "
                            "Must be a class or a str.".format(type(classid)))
        if classid not in Configurator.registry:
            raise AttributeError(
                "Argument is not a class or classid of registered base class")
        self.baseRegistry = Configurator.registry[classid]
        assert classid in self.baseRegistry

        
    def __new__(cls, *args, **kwargs):
        if not cls._instance:
            cls._instance = super(Configurator, cls).__new__(cls)
        return cls._instance
    
    def registerClass(self, derived):
        self.baseRegistry[derived.__classid__] = derived
        return self
        
    def getSchema(self, classid, rules = AssemblyRules(AccessType(READ | WRITE | INIT))):
        '''
        Get schema for class with "classid" derived from base class given to constructor using assembly "rules"
        Example:
                schema = Configurator(Shape).getSchema("Rectangle")
        '''
        if isinstance(classid, type):
            classid = classid.__classid__
        if not isinstance(classid, str):
            raise TypeError("The first argument type '{}' is not allowed. "
                            "Must be a type or str.".format(type(classid)))
        if classid not in self.baseRegistry:
            raise AttributeError("Class Id '{}' not found in the base registry"
                                 .format(classid))
        Derived = self.baseRegistry[classid]   # userclass -> Derived

        # building list of classes in inheritance order from bases to the last derived
        def inheritanceChain(c, bases_id, clist):
            if not isinstance(c, type):
                return
            if c.__classid__ not in bases_id:
                for x in c.__bases__:
                    inheritanceChain(x, bases_id, clist)
            if c not in clist:
                clist.append(c)
                
        clist = []
        inheritanceChain(Derived, Derived.__bases_classid__, clist)
        # clist contains list of classes in inheritance order
        schema = Schema(classid, rules)
        for theClass in clist:
            try:
                if hasattr(theClass, "expectedParameters"):
                    theClass.expectedParameters(schema) # fill schema in order from base to derived
            except AttributeError as e:
                print "Exception while adding expected parameters for class %r: %r" % (theClass.__name__, e)
        return schema
    
    
    def create(self, *args):
        '''
        The factory method to create the instance of class with "classId" that inherits from base class given to constructor using "input" configuration.
        The last argument is a flag to determine if the input configuration should be validated.
        Example:
                instance = Configurator(Shape).create("EditableCircle", Hash("radius", 12.345))
                
        The factory method to create instance of class that inherits from base class given to constructor using input "configuration".
        The configuration should have "classId" of class to be created as a root element.  The last argument is a flag to determine
        if the input "configuration" should be validated.
        Example:
                configuration = Hash("EditableCircle.radius", 12.345)
                instance = Configurator(Shape).create(configuration)
        '''
        if len(args) == 3:
            classid = args[0]
            configuration = args[1]
            validation = args[2]
        elif len(args) == 2:
            if type(args[1]) == bool:
                configuration = args[0]
                validation = args[1]
                classid = configuration.keys()[0]
                configuration = configuration[classid]
            else:
                classid = args[0]
                configuration = args[1]
                validation = True
        elif len(args) == 1:
            configuration = args[0]
            validation = True
            classid = configuration.keys()[0]
            configuration = configuration[classid]
        else:
            raise TypeError("Wrong number of arguments and/or their types")
        if isinstance(classid, type):
            classid = classid.__classid__
        if not isinstance(classid, basestring):
            raise TypeError(
                "First argument 'classid' must be a python string type")
        if classid not in self.baseRegistry:
            raise AttributeError("Unknown classid '{}' in base registry".
                                 format(classid))
        Derived = self.baseRegistry[classid]
        schema = Configurator(Derived.__base_classid__).getSchema(classid)
        if not validation:
            return Derived(configuration)
        validated = Hash()
        validator = Validator()
        try:
            validated = validator.validate(schema, configuration)
        except RuntimeError as e:
            raise RuntimeError("Validation Exception: " + str(e))
        return Derived(validated)
    
    
    def createNode(self, nodename, classid, configuration, validation = True):
        '''
        The helper method to create instance of class specified by "classId" and derived from class given to constructor using
        sub-configuration specified by "nodeName" which has to be a part of input "configuration".
        The last argument is a flag to determine if the input "configuration" should be validated.
        '''
        if isinstance(classid, type):
            classid = classid.__classid__
        if not isinstance(classid, str):
            raise TypeError(
                "Second argument 'classid' must be a python string type")
        if nodename in configuration:
            return self.create(classid, configuration[nodename], validation)
        raise AttributeError(
            'Given nodeName "{}" is not part of input configuration'.
            format(nodename))

    def createChoice(self, choicename, configuration, validation = True):
        '''
        The helper method to create the instance of class derived from base class given to constructor using "choiceName" and
        input "configuration".  The last argument is a flag to determine if the input configuration should be validated.
        '''
        return self.create(configuration[choicename], validation)
       
    def createList(self, listname, input, validation = True):
        '''
        The helper method to create the list of instances of classes derived from base class given to constructor using "listName"
        used as a key to the list and "input" configuration.  The last argument is a flag to determine if the input configuration
        should be validated.
        '''
        if listname not in input:
            raise AttributeError(
                'Given list name "{}" is not a part of input configuration'.
                format(listname))
        instances = []
        for hash in input[listname]:
            instances.append(create(hash, validation))
        return instances
            
    def getRegisteredClasses(self):
        '''
        Returns list of "classid"'s for all registered classes derived from base class given to constructor.
        '''
        return self.baseRegistry.keys()
    
    @staticmethod
    def getRegisteredBaseClasses():
        '''
        Returns all classid's of base classes registered in Configurator.
        '''
        return Configurator.registry.keys()
